### PR TITLE
ADEN-5713 Add support for A9 video

### PR DESCRIPTION
**Pull request by v3rth**

### CODE REVIEW

File: extensions/wikia/AdEngine/js/video/ooyalaAdSetProvider.js [hunk 1]
```diff
@@ -28,16 +28,17 @@ define('ext.wikia.adEngine.video.ooyalaAdSetProvider', [
 			}
 		};
 
-	function generateSet(ad, rv, correlator, videoInfo) {
+	function generateSet(ad, rv, correlator, videoInfo, bidParams) {
```
Comment (WikiaTech) [inline extensions/wikia/AdEngine/js/video/ooyalaAdSetProvider.js:32]: MAJOR Function has 5 parameters which is greater than 4 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-js-rules%3AAmdExcludedFunctionParameterLimit)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 2]
```diff
@@ -83,22 +85,34 @@ require([
...
+			}
+
 			if (ooyalaAdSetProvider.canShowAds()) {
-				options.adSet = ooyalaAdSetProvider.get(1, correlator, {
-					contentSourceId: videoData.dfpContentSourceId,
-					videoId: videoId
-				});
 				options.replayAds = ooyalaAdSetProvider.adsCanBePlayedOnNextVideoViews();
+
+				a9.waitForResponse()
```
Comment (fraszczakszymon) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js]: What if it's undefined?

File: extensions/wikia/AdEngine/js/lookup/lookupFactory.js [hunk 3]
```diff
@@ -122,6 +122,25 @@ define('ext.wikia.adEngine.lookup.lookupFactory', [
...
+			}, milisToTimeout);
+		}
+
+		function waitForOrTimeout(f, milisToTimeout) {
+			milisToTimeout = milisToTimeout || 2000;
+			var timeout = new Promise(function (resolve, reject) {
+					setTimeout(reject, milisToTimeout);
+				});
+
+			return Promise.race([new Promise(f), timeout]);
```
Comment (kvas-damian) [inline extensions/wikia/AdEngine/js/lookup/lookupFactory.js]: Promise is not supported by IE11 :trollface:

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 4]
```diff
@@ -83,22 +85,34 @@ require([
...
+						options.adSet = ooyalaAdSetProvider.get(1, correlator, {
+							contentSourceId: videoData.dfpContentSourceId,
+							videoId: videoId
+						}, additionalSlotParams);
+
+						ooyalaVideoController = OoyalaPlayer.initHTML5Player(ooyalaVideoElementId, options, onCreate);
+					});
+
 			} else {
 				playerTrackerParams.adProduct = 'featured-video-no-preroll';
```
Comment (kvas-damian) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js]: Is it ok that we set params for tracking after we called `track()`?

File: extensions/wikia/AdEngine/js/lookup/lookupFactory.js [hunk 5]
```diff
@@ -122,6 +122,25 @@ define('ext.wikia.adEngine.lookup.lookupFactory', [
 			});
 		}
 
+		function waitForResponse(milisToTimeout) {
+			return waitForOrTimeout(function (resolve) {
```
Comment (kvas-damian) [inline extensions/wikia/AdEngine/js/lookup/lookupFactory.js]: wait for what?

File: extensions/wikia/AdEngine/js/lookup/lookupFactory.js [hunk 6]
```diff
@@ -122,6 +122,25 @@ define('ext.wikia.adEngine.lookup.lookupFactory', [
...
+				if (hasResponse()) {
+					resolve();
+				} else {
+					addResponseListener(resolve);
+				}
+			}, milisToTimeout);
+		}
+
+		function waitForOrTimeout(f, milisToTimeout) {
```
Comment (kvas-damian) [inline extensions/wikia/AdEngine/js/lookup/lookupFactory.js]: should it be part of `lookupFactory`?

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js (hunk 2 above)
Comment (v3rth) [reply extensions/wikia/ArticleVideo/scripts/article-featured-video.js]: :boom:

File: extensions/wikia/AdEngine/js/lookup/lookupFactory.js (hunk 3 above)
Comment (v3rth) [reply extensions/wikia/AdEngine/js/lookup/lookupFactory.js]: http://community.wikia.com/wiki/Help:Supported_browsers :trollface:

File: resources/wikia/modules/promise.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+/**
+ * AMD module wrapping Promise global object
+ */
+define('wikia.promise', function () {
+	'use strict';
+
+	var P = window.Promise;
```
Comment (kvas-damian) [inline resources/wikia/modules/promise.js]: `wikia.window`?

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 8]
```diff
@@ -84,22 +86,42 @@ require([
...
 			if (ooyalaAdSetProvider.canShowAds()) {
-				options.adSet = ooyalaAdSetProvider.get(1, correlator, {
-					contentSourceId: videoData.dfpContentSourceId,
-					videoId: videoId
-				});
 				options.replayAds = ooyalaAdSetProvider.adsCanBePlayedOnNextVideoViews();
+
+				if (a9) {
+					a9.waitForResponse()
+						.then(function () { return a9.getSlotParams('FEATURED'); })
```
Comment (kvas-damian) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js]: Are you sure those slot params are enough for A9 video targeting? I see documentation is advising us to use `qsParams`.

Comment (v3rth) [reply extensions/wikia/ArticleVideo/scripts/article-featured-video.js]: There is no `qsParams` parameter :disappointed: 
image

Comment (kvas-damian) [reply extensions/wikia/ArticleVideo/scripts/article-featured-video.js]: We talked offline - there is `qsParams` attribute ;)
Code is fixed.
For now, let's use `getSlotParams()` and see how it works while QA.

**Comment by kvas-damian** [comment]
Please make sure kikimora understands a9 video. New column or put data as existing bidder column

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [adContext.js#L78](https://github.com/Wikia/app/blob/bddfacce66c8579cbf15258c466457dff4f82f9d/extensions/wikia/AdEngine/js/context/adContext.js#L78): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1067)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 9]
```diff
@@ -84,22 +86,42 @@ require([
...
 
 			if (ooyalaAdSetProvider.canShowAds()) {
-				options.adSet = ooyalaAdSetProvider.get(1, correlator, {
-					contentSourceId: videoData.dfpContentSourceId,
-					videoId: videoId
-				});
 				options.replayAds = ooyalaAdSetProvider.adsCanBePlayedOnNextVideoViews();
+
+				if (a9) {
+					a9.waitForResponse()
```
Comment (kvas-damian) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js]: What if `VideoBidder` is disabled?

File: extensions/wikia/AdEngine/js/lookup/a9.js [hunk 10]
```diff
@@ -22,30 +23,44 @@ define('ext.wikia.adEngine.lookup.a9', [
...
-			a9Script.src = '//c.amazon-adsystem.com/aax2/apstag.js';
+		node.parentNode.insertBefore(a9Script, node);
+	}
 
-			node.parentNode.insertBefore(a9Script, node);
+	function isVideoBidderEnabled() {
+		return adContext &&
+			adContext.getContext() &&
+			adContext.getContext().bidders &&
+			adContext.getContext().bidders.a9Video
```
Comment (kvas-damian) [inline extensions/wikia/AdEngine/js/lookup/a9.js]: Can we avoid calling `getContext()` 3 times in the same method? :)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [adContext.js#L78](https://github.com/Wikia/app/blob/576253f630d4f42bcd144f3ee627e1f851f70ddc/extensions/wikia/AdEngine/js/context/adContext.js#L78): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1067)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 11]
```diff
@@ -84,22 +89,42 @@ require([
...
-				options.adSet = ooyalaAdSetProvider.get(1, correlator, {
-					contentSourceId: videoData.dfpContentSourceId,
-					videoId: videoId
-				});
 				options.replayAds = ooyalaAdSetProvider.adsCanBePlayedOnNextVideoViews();
+
+				if (a9 && context.bidders && context.bidders.a9Video) {
+					a9.waitForResponse()
+						.then(function () { return a9.getSlotParams('FEATURED'); })
+						.catch(function () { return {}; })
```
Comment (nandy-andy) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js:102]: Should we at least log it to the console? Ideally, we could send it to Kibana ;)

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 12]
```diff
@@ -84,22 +89,42 @@ require([
...
 			if (playerTracker) {
 				playerTracker.track(playerTrackerParams, 'init');
 			}
 
 			ooyalaVideoController = OoyalaPlayer.initHTML5Player(ooyalaVideoElementId, options, onCreate);
+		}
 
-			document.addEventListener('visibilitychange', handleTabChange);
+		function setupFirstAdSet(additionalSlotParams) {
+			return ooyalaAdSetProvider.get(1, correlator, {
```
Comment (nandy-andy) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js:129]: What does `1` represent here? :)

Comment (v3rth) [reply extensions/wikia/ArticleVideo/scripts/article-featured-video.js:129]: it's 1-st video play

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js (hunk 11 above)
Comment (v3rth) [reply extensions/wikia/ArticleVideo/scripts/article-featured-video.js:102]: I think there is no good place for storing bidder errors and no one want to check those errors :)
Main function of this catch is handling timeouts. As far as I understand, timeout for bidders should be logged anyway

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 13]
```diff
@@ -7,8 +7,10 @@ require([
 	'wikia.geo',
 	'wikia.instantGlobals',
 	'wikia.articleVideo.videoFeedbackBox',
+	'ext.wikia.adEngine.adContext',
 	require.optional('ext.wikia.adEngine.video.player.ooyala.ooyalaTracker'),
-	require.optional('ext.wikia.adEngine.video.ooyalaAdSetProvider')
+	require.optional('ext.wikia.adEngine.video.ooyalaAdSetProvider'),
+	require.optional('ext.wikia.adEngine.lookup.a9')
```
Comment (dianafa) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js]: alphabetical order :))

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 14]
```diff
@@ -84,22 +89,42 @@ require([
...
+					a9.waitForResponse()
+						.then(function () { return a9.getSlotParams('FEATURED'); })
+						.catch(function () { return {}; })
+						.then(function (additionalSlotParams) { // finally
+							options.adSet = setupFirstAdSet(additionalSlotParams);
+							initPlayerWithTracking(options, onCreate);
+						});
+				} else {
+					options.adSet = setupFirstAdSet();
+					initPlayerWithTracking(options, onCreate);
```
Comment (dianafa) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js:109]: those 2 lines are almost the same as those from l99, what do you think about extracting it to some function?

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js [hunk 15]
```diff
@@ -84,22 +89,42 @@ require([
 				};
 
 			if (ooyalaAdSetProvider.canShowAds()) {
-				options.adSet = ooyalaAdSetProvider.get(1, correlator, {
-					contentSourceId: videoData.dfpContentSourceId,
-					videoId: videoId
-				});
 				options.replayAds = ooyalaAdSetProvider.adsCanBePlayedOnNextVideoViews();
+
+				if (a9 && context.bidders && context.bidders.a9Video) {
```
Comment (dianafa) [inline extensions/wikia/ArticleVideo/scripts/article-featured-video.js:99]: it's the same check as in `isVideoBidderEnabled` - could we reuse that function?

File: includes/wikia/DefaultSettings.php [hunk 16]
```diff
@@ -1199,6 +1199,13 @@
  */
 $wgAmazonMatchCountriesMobile = null;
 
+/**
+ * @name wgAdDriverA9VideoBidderCountries
+ * List of countries where A9 video bidding platform is enabled.
+ * It won't work if A9 display bidder isn't enabled
+ * ONLY UPDATE THROUGH WIKI FACTORY ON COMMUNITY - it's an instant global.
+ */
+$wgAdDriverA9VideoBidderCountries = null;
```
Comment (dianafa) [inline includes/wikia/DefaultSettings.php]: could we set it to [] ?

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js (hunk 15 above)
Comment (v3rth) [reply extensions/wikia/ArticleVideo/scripts/article-featured-video.js:99]: No, we can't, it's behind factory interface :(

File: extensions/wikia/ArticleVideo/scripts/article-featured-video.js (hunk 14 above)
Comment (v3rth) [reply extensions/wikia/ArticleVideo/scripts/article-featured-video.js:109]: I considered that, but I think it's cleaner to keep it like this, in other case state of `option` object will be modified from external function

**Comment by WikiaTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [adContext.js#L78](https://github.com/Wikia/app/blob/d6a720e9457b78263beb130ce6d4786ab92cf64c/extensions/wikia/AdEngine/js/context/adContext.js#L78): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1067)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by dianafa** [comment]
LGTM

File: extensions/wikia/AdEngine/js/video/ooyalaAdSetProvider.js [hunk 17]
```diff
@@ -72,20 +73,21 @@ define('ext.wikia.adEngine.video.ooyalaAdSetProvider', [
 		return !isReplay || (isReplay && cappingAllowsToShowNextVideo(videoDepth, adsFrequency));
 	}
 
-	function get(videoDepth, correlator, videoInfo) {
+	function get(videoDepth, correlator, videoInfo, bidParams) {
```
Comment (WikiaTech) [inline extensions/wikia/AdEngine/js/video/ooyalaAdSetProvider.js:84]: MAJOR Function has a complexity of 11 which is greater than 10 authorized. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AFunctionComplexity)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [adContext.js#L78](https://github.com/Wikia/app/blob/f4e04e502fb8c7d92ef098ea233bd3308634ca94/extensions/wikia/AdEngine/js/context/adContext.js#L78): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1067)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Review by kvas-damian — APPROVED** [review]

**Review by fraszczakszymon — APPROVED** [review]

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [adContext.js#L78](https://github.com/Wikia/app/blob/1e07726440cae688ab670717fb86120d7b8f64cd/extensions/wikia/AdEngine/js/context/adContext.js#L78): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1067)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by WikiaTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [adContext.js#L78](https://github.com/Wikia/app/blob/b3851ae43068208295f97533c919cb63811c6413/extensions/wikia/AdEngine/js/context/adContext.js#L78): Reduce the number of conditional operators (4) used in the expression (maximum allowed 3). [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=javascript%3AS1067)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**